### DISCUSSION TITLE
A more sustainable strategy for cache storage

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#1875] [General]
Perhaps a naive question. Given 1) the difficulties encountered with this cache export feature, 2) this is one of the easy ones (users will want to export cache to many various targets, most of which are not supported by buildkit at all)...

Perhaps we should start discussing an alternative design that is more sustainable long term (while we continue working on fixing this).

My observations:

- Buildkit local exporter is very mature, and works everywhere.
- The local exporter produces mostly immutable content-addressed blobs, and a little metadata.
- If dagger always exported its cache locally, then synchronized the local cache to a pluggable storage backend, it would be very easy to export cache in that way, including to GHA cache.
- *importing* cache would be more difficult per @aluzzardi (how do you know which blobs to download) but perhaps tackling that problem, although difficult, is a better tradeoff than climbing a new mountain for each new export target we need buildkit to support? Given the current difficulties with one of the *best-supported* targets, that seems unsustainable. I think we need a plan B.

My 2c.

__Originally posted by @shykes in https://github.com/dagger/dagger/issues/1020#issuecomment-1067072805__

**Comment by shykes** [comment]
paging @grouville @aluzzardi @gerhard @TomChv

**Comment by TomChv** [comment]
Just a few reaction

> Buildkit local exporter is very mature, and works everywhere

Currently local cache doesn't work on Dagger, I'm still working on it.

Otherwise I agree with your proposal, I think with time we will have two solutions : 
1. Add new storage to buildkit to match our needs (we are more dependant to buildkit)
2. Find a generic way that works in any cases, any environments 

One of my concern is when we will need to integrate Gitlab runner, or Tekton or anything, it can become really painful to handle cache directly with Buildkit. 
That's sure that will be really easier with homemade solution

**Comment by gerhard** [comment]
While I have not read all the content above, I wanted to mention about the good progress that @TomChv made with @crazy-max today: https://github.com/dagger/dagger/pull/1731#issuecomment-1068271318

**Comment by aluzzardi** [comment]
@TomChv made a lot of progress and the fix is near. The problem was with the entire cache system, not just GHA.

There are several issues with buildkit caching, including the fact that its memory is short (only the last build, it's not cumulative).

I think the long term solution is to revamp it, either by contributing upstream a more modular approach (had a great conversation recently where someone mentioned the idea of having external cache backends in containers, similar to how frontends are handled) OR by moving the cache to a lower layer (e.g. containerd snapshots).

I think using the `local` backend and then handling things out of bound is a bit challenging both for import and export because we don't know what buildkit needs ahead of time, nor what exactly changed/was unchanged, so we'd have to export everything.

**Comment by aluzzardi** [comment]
Did a bit of research using our own `ci` plan as a test.

Local cache takes about ~30 seconds after `do` completed to export ~800MB of data, **every time**.

```console
$ time dagger do build
dagger do build  3.56s user 1.81s system 104% cpu 5.118 total

$ time dagger do build --cache-to type=local,dest=cache1
dagger do build --cache-to type=local,dest=cache1  15.17s user 6.45s system 69% cpu 31.202 total

$ time dagger do build --cache-to type=local,dest=cache2
dagger do build --cache-to type=local,dest=cache2  16.46s user 7.02s system 55% cpu 42.397 total

$ du -sh cache1
826M	cache1
```

Comparatively, an export to registry only has a few seconds of overhead on the first run (NOTE: local registry) but has NO overhead the second run (since no bits need exporting):

```console
$ time dagger do build --cache-to type=registry,ref=localhost:5555/cachetest
dagger do build --cache-to type=registry,ref=localhost:5555/cachetest  3.61s user 2.28s system 82% cpu 7.106 total

$ time dagger do build --cache-to type=registry,ref=localhost:5555/cachetest
dagger do build --cache-to type=registry,ref=localhost:5555/cachetest  3.58s user 2.51s system 149% cpu 4.072 total
```

**Comment by shykes** [comment]
> Local cache takes about ~30 seconds after `do` completed to export ~800MB of data, **every time**.
> 
> Comparatively, an export to registry only has a few seconds of overhead on the first run (NOTE: local registry) but has NO overhead the second run (since no bits need exporting):

*warning: thinking out loud*

That certainly makes always-local cache a less attractive strategy. Perhaps we should consider always-registry? That does have its own problems: is there always a suitable registry that dagger can use without requiring heavy-duty infrastructure work? And if not, can dagger bootstrap its own, similar to buildkit? But then how to persist its state?
